pkg/ebpf/uprobes/doc.go

```diff
@@ -0,0 +1,58 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+/*
+Package uprobes contains methods to help handling the attachment of uprobes to
+userspace programs
+
+The main type for this package is the UprobeAttacher type, created with
+NewUprobeAttacher. The main configuration it requires is a list of rules that
+define how to match the possible targets (shared libraries and/or executables)
+and which probes to attach to them. Example usage:
+
+	connectProbeID := manager.ProbeIdentificationPair{EBPFFuncName: "uprobe__SSL_connect"}
+	mainProbeID := manager.ProbeIdentificationPair{EBPFFuncName: "uprobe__main"}
+
+	mgr := manager.Manager{}
+
+	attacherCfg := &AttacherConfig{
+		Rules: []*AttachRule{
+			{
+				LibraryNameRegex: regexp.MustCompile(`libssl.so`),
+				Targets:          AttachToSharedLibraries,
+				ProbesSelector: []manager.ProbesSelector{
+					&manager.ProbeSelector{ProbeIdentificationPair: connectProbeID},
+				},
+			},
+			{
+				Targets: AttachToExecutable,
+				ProbesSelector: []manager.ProbesSelector{
+					&manager.ProbeSelector{ProbeIdentificationPair: mainProbeID},
+				},
+			},
+		},
+		ExcludeTargets: ExcludeInternal | ExcludeSelf,
+		EbpfConfig:     ebpfCfg,
+	}
+
+	ua, err := NewUprobeAttacher("test", attacherCfg, &mgr, callback, &NativeBinaryInspector{})
+	ua.Start()
+
+Once started, the attacher monitors new processes and `open` calls for new
+shared libraries. For the first task it uses pkg/process/monitor/ProcessMonitor,
+and for the second it uses the shared-libraries program in
+pkg/network/usm/sharedlibraries.
+
+# Notes and things to take into account
+
+  - When adding new probes, be sure to add the corresponding code to
+    match the libraries in
+    pkg/network/ebpf/c/shared-libraries/probes.h:do_sys_open_helper_exit, as an
+    initial filtering is performed there.
+
+  - If multiple rules match a binary file, and we fail to attach the required probes for one of them,
+    the whole attach operation will be considered as failed, and the probes will be detached.
```

Comment:
for shared libraries, we have a "best effort" approach for some of the probes (like `SSL_readex` in openssl). How is it represented in uprobe attacher?

Reply:
As these are libraries, you'd have a rule that matches the library file, and then a `ProbesSelector` that selects the probes to use, with `manager.BestEffort/manager.AllOf` to signal which ones are optional and which aren't. The idea is to fully reuse the selections you have such as [`openSSLProbes`](https://github.com/DataDog/datadog-agent/blob/guillermo.julian/uprobe-attacher/pkg/network/usm/ebpf_ssl.go#L74).

That comment refers more to what to do when two rules match the same executable/library. I think the optimal way to do this is to design the rules so that only one rule might match each file, but I'm not sure whether it's worth it to enforce that in code.